Initial Python QPS Worker

kpayson64 opened the pull request:
This is the initial python qps worker.  It currently only supports a SYNC server with proto (non-raw) messages.  It doesn't have hooks into the Jenkins test, but can be run with ./run_test.py (options)

jtattermusch commented on src/python/grpcio/tests/qps/intervals.py:
```diff
@@ -0,0 +1,42 @@
+ while True:
+   yield random.expovariate(alpha)
+
+def uniform(low, high):
```
FYI, uniform and deterministic payload types will be removed soon, so there's no need to implement those.

jtattermusch commented on src/python/grpcio/tests/qps/qps_worker.py:
```diff
@@ -0,0 +1,58 @@
+  server.add_insecure_port('[::]:{}'.format(port))
+  server.start()
+  servicer.wait_for_quit()
+  time.sleep(2)
```
why the sleep?

kpayson64 replied:
servicer.wait_for_quit() returns before the stub.Quit() call is done
I tried server.stop(5), server.stop(5).wait(), sever.stop(0).wait(), but none of these variants let the sever exit cleanly, I end up with some variant of   
D0418 17:59:28.805390478   80996 chttp2_transport.c:698]     got goaway [0]: 53 65 72 76 65 72 20 73 68 75 74 64 6f 77 6e 'Server shutdown'

The sleep is a hacky way of a clean exit

jtattermusch commented on src/python/grpcio/tests/qps/worker_server.py:
```diff
@@ -0,0 +1,172 @@
+import src.proto.grpc.testing.services_pb2 as services_pb2
+
+""" Python Worker Server implementation.  Responsible for running servers 
+and clients on demend.
```
s/demend/demand/

jtattermusch commented:
I don't think for python we check in the generated _pb2 files into the repository (they're generated on demand), so they probably need to be removed from this PR.

jtattermusch commented on src/python/grpcio/tests/qps/src/proto/grpc/testing/.gitignore:
```diff
@@ -0,0 +1 @@
+!*_pb2.py
```
The _pb2.py ignore is there because we don't want to check in generated files for python. Please revert this change.

jtattermusch commented:
Ad run_test.py script:

We actually already have two cmdline utils to run the performance tests (written in C++):  qps_driver and qps_json_driver, so having an extra test driver written in python seems not to be useful. Also, we're probably trying to abandon the qps_driver approach that uses lots of cmdline arguments to configure the tests and using JSON to configure the scenarios seems to be the way to go. 

That said, I think the qps_driver written in C++ is a bit of an overkill and in the future we might want to replace the qps_driver with something written in a more high-level language (perhaps python) - but there would be some extra challenges involved, so for now we will probably stick with the C++ qps_json_driver.
Can we remove run_test.py from this PR (but let's keep the code in case we need it) and instead test the python workers with qps_json_driver?

Btw, integrating python with the rest of the performance testing framework should be easy enough - 
see https://github.com/grpc/grpc/pull/6204/files which is adding ruby integration (we don't need to do integration in this PR, but that's where we are aiming).

kpayson64 commented:
I've updated the tests to be run using tools/run_tests/run_performance_tests.py -l python.  I've also removed the _pb2 files from the commit.  This branch was too far behind master, and was missing some stuff around qps_json_driver, so I've moved this to #6216
